### PR TITLE
Version3 listener unstable

### PR DESCRIPTION
**Pull request by jasonyurs**
Stabilized StreamReader abstraction changes

### CODE REVIEW

File: listener/listener_test.go [hunk 1]
```diff
@@ -101,36 +54,58 @@ func TestListener(t *testing.T) {
 		So(err, ShouldBeNil)
 		So(len(shards), ShouldEqual, 1)
 
-		l, err := NewListener(stream, shards[0], func(c *Config) {
+		p, err := producer.NewProducer(func(c *producer.Config) {
```
Comment (twang-rs) [inline listener/listener_test.go]: If we create a producer, we don't need putRecord above.

It may be simpler to keep putRecord instead of creating a Producer, as producers have an inherent delay due to their batch sizes.

File: producer/producer.go [hunk 2]
```diff
@@ -48,8 +48,8 @@ type Producer struct {
 	concurrencySem chan Empty
 	pipeOfDeath    chan Empty
 	outstanding    int
-	shutdownCond   sync.Cond
-	producerWg     sync.WaitGroup
+	shutdownCond   *sync.Cond
```
Comment (twang-rs) [inline producer/producer.go]: Were these changes actually necessary?

File: circle.yml [hunk 3]
```diff
@@ -13,7 +13,7 @@ dependencies:
     - go get honnef.co/go/tools/cmd/staticcheck
     - go get honnef.co/go/tools/cmd/gosimple
     - go get -u github.com/golang/lint/golint
-    - docker run -d -t -p 4567:4567 dlsniper/kinesalite:1.8.0
+    - docker run -d -t -p 4567:4567 instructure/kinesalite:latest
```
Comment (jasonyurs) [inline circle.yml]: Recommendation by @DocMerlin.  Not sure if it is absolutely necessary though.

File: producer/producer.go (hunk 2 above)
Comment (jasonyurs) [inline producer/producer.go]: These changes were causing failures in either go lint, go vet or go simple.

File: listener/kclreader.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+				}
+
+				// Unmarshall the decoded message
+				msgs := []message.Message{}
+				err = json.Unmarshal(decodedMsg, msgs)
+				if err != nil {
+					r.listener.LogError(err)
+					return err
+				}
+
```
Comment (DocMerlin) [inline listener/kclreader.go]: you need to unmarshal then base64 decode.

File: multilang/action.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	Checkpoint ActionType = "checkpoint"
+	Shutdown ActionType = "shutdown"
+	Status ActionType = "status"
+)
+
+// ActionMessage is a struct used to marshal / unmarshall KCL Multilang protocol action messages
+type ActionMessage struct {
+	Action      ActionType `json:"action"`
+	ShardID     string `json:"shardId,omitempty"`
+	Records     string `json:"records,omitempty"`
```
Comment (DocMerlin) [inline multilang/action.go]: this should have a richer type, as in the input it is a json array of records.
